### PR TITLE
CP-48027: Corosync upgrade add `cluster_stack_version` datamodel change

### PR DESCRIPTION
**Pull request by Vincent-lau**
This is the core part of the corosync upgrade, mainly deal with the `cluster_stack_version` datamodel field, and the switching logic for corosync binary and libraries.

The multi-link support will be done in a separte PR.

Best reviewed by commit

### CODE REVIEW

**Comment by lindig** [comment]
Something is wrong here? 5000+ files changed?

**Comment by Vincent-lau** [comment]
> Something is wrong here? 5000+ files changed? You might want to split automated tests into a different commit so we can ignore them.

There are autogenerated test files due to the interface change. I could do a separate commit, but that would mean some of the commits will fail the test though. But it looks like this is probably the only way to make the CI happy.

**Comment by psafont** [comment]
The review dog thing failing is really not good :/

File: .github/workflows/main.yml
Comment (psafont) [inline .github/workflows/main.yml:101]: Now it won't block when there are actual errors, but at least will still make the annotations visible

File: .github/workflows/main.yml
Comment (psafont) [inline .github/workflows/main.yml:115]: Same here

File: ocaml/idl/datamodel_cluster.ml
Comment (minglumlu) [inline ocaml/idl/datamodel_cluster.ml:198]: Would it be better to following the way of the "cluster_stack" field, making the default value from `Constants.default_smapi_v3_cluster_stack_version`?

File: ocaml/tests/common/test_common.ml
Comment (minglumlu) [inline ocaml/tests/common/test_common.ml:638]: Why is it `3` here, but it's `2` in datamodel?

File: ocaml/idl/datamodel_cluster.ml
Comment (minglumlu) [reply ocaml/idl/datamodel_cluster.ml:198]: Hopefully the version will never be something like `2.5`.

File: ocaml/xapi/xapi_cluster.ml
Comment (minglumlu) [inline ocaml/xapi/xapi_cluster.ml]: Would this be just the default value?

File: ocaml/xapi/xapi_cluster_host.ml
Comment (minglumlu) [inline ocaml/xapi/xapi_cluster_host.ml]: Splitting the `stack` and `version` seems annoying to me.
Given the following has been defined:
```
type t = Corosync2 | Corosync3
```
How about encoding the version into the type values? E.g.
```
module Cluster_stack = struct
  let of_string = function
    | "corosync" -> Corosync2
    | "corosync3" -> Corosync3
    | _ -> ...
end
```

File: ocaml/xapi/xapi_cluster_helpers.ml
Comment (minglumlu) [inline ocaml/xapi/xapi_cluster_helpers.ml]: There is a `restrict_corosync` already. Are we going to separate features for clustering?

File: ocaml/xapi/xapi_clustering.ml
Comment (minglumlu) [inline ocaml/xapi/xapi_clustering.ml]: Ah. I see. This feature flag is only for the function `switching an existing cluseter from corosync2 to corosync3`. A user can still create a new cluster with corosync3 without caring about this new feature flag.

File: ocaml/idl/datamodel_cluster.ml
Comment (Vincent-lau) [reply ocaml/idl/datamodel_cluster.ml:198]: sounds good, avoid magic constants

File: ocaml/tests/common/test_common.ml
Comment (Vincent-lau) [reply ocaml/tests/common/test_common.ml:638]: The `2` in the datamodel is for backwards compatibility reasons, when we upgrade from an old xapi db which does not have the field, but has a cluster object, we can default that field to 2, because upgrading from an old xapi necessarily means corosync has version 2.

The 3, OTOH, is used as the "default" version when we _create_ a new cluster, since we want to enforce that whenever the user is creating a new cluster in XS 9. But in this case it does not really matter because it is not tested :)

I think this can be confusing so maybe I will add a clarification in the datamodel

File: ocaml/idl/datamodel_cluster.ml
Comment (Vincent-lau) [reply ocaml/idl/datamodel_cluster.ml:198]: This can be counter-intuitive, I will add a comment in the doc.

File: ocaml/xapi/xapi_cluster.ml
Comment (Vincent-lau) [reply ocaml/xapi/xapi_cluster.ml]: see comments above

File: ocaml/xapi/xapi_clustering.ml
Comment (Vincent-lau) [reply ocaml/xapi/xapi_clustering.ml]: Yep! This feature is to restrict the upgrade process, not restricting the feature itself. A user would be able to create a corosync2 cluster as usual in XS 8.

File: ocaml/xapi/xapi_cluster_helpers.ml
Comment (Vincent-lau) [reply ocaml/xapi/xapi_cluster_helpers.ml]: Yes see comments below

**Comment by psafont** [comment]
Can the conflicts be solved first?

File: ocaml/idl/datamodel_cluster.ml
Comment (robhoes) [reply ocaml/idl/datamodel_cluster.ml:198]: >We will always create a corosync3 cluster when the user does not specify anything

But only once we are ready for Corosync 3, i.e. not yet.

File: ocaml/xapi/xapi_clustering.ml
Comment (robhoes) [inline ocaml/xapi/xapi_clustering.ml]: I think switching the corosync version and getting the current corosync version is a job for the clustering daemon. Can we add functions to the cluster interface for this?

File: ocaml/xapi-idl/cluster/cluster_interface.ml
Comment (robhoes) [inline ocaml/xapi-idl/cluster/cluster_interface.ml:364]: I think this string should match the function name, so no `_version` at the end. Also, the other ones above use hyphens. I think it is used in the CLI that is generated from this.

File: ocaml/xapi-idl/cluster/cluster_interface.ml
Comment (robhoes) [inline ocaml/xapi-idl/cluster/cluster_interface.ml:363]: Could we use `Cluster_stack.t`?

File: ocaml/xapi/xapi_cluster.ml
Comment (robhoes) [inline ocaml/xapi/xapi_cluster.ml]: It would be nicer to use `Cluster_stack.t` rather than all those string concatenations here and below.

File: ocaml/xapi/xapi_clustering.ml
Comment (robhoes) [inline ocaml/xapi/xapi_clustering.ml]: Can we use
```
let dbg = Context.string_of_task_and_tracing __context in
```
and `dbg` for that debuginfo parameter?

File: ocaml/xapi-idl/cluster/cluster_interface.ml
Comment (Vincent-lau) [reply ocaml/xapi-idl/cluster/cluster_interface.ml]: Yes, good point. Fixed

File: ocaml/xapi/xapi_clustering.ml
Comment (Vincent-lau) [reply ocaml/xapi/xapi_clustering.ml]: Yep!

**Review by robhoes — APPROVED** [review]

**Review by psafont — APPROVED** [review]

**Comment by Vincent-lau** [comment]
hold off merging as there are non-trivial changes after the review. I would like to do some xenrt testing before merging it.

**Comment by Vincent-lau** [comment]
test looks good, merge it now

**Pull request merged**
